### DISCUSSION TITLE
Include certain hidden files in templates (e.g. based on .quartoignore)

### DISCUSSION THREAD

**Discussion by skwde** [quarto-dev/quarto-cli#9351] [Q&A]
### Description

As mentioned here https://quarto.org/docs/extensions/starter-templates.html#creating-a-template all hidden files / directories are ignored. 

It would be nice to have a way to allow certain hidden files / directories.
This would allow to also bundle common project setup with the template like lint config files or a `.devcontainer` 

I tried using a `.quartoignore` or `.gitignore` like

```
!.devcontainer/
!.devcontainer/*
!.pre-commit-config.yaml
```

but this didn't help.

**Comment by mcanouil** [comment]
Use `format-resources` and/or `resources` to add resources depending on what your extension is.
It should work, if not please share a small Git repository which reproduce the issue.

**Reply by skwde** [reply]
Please excuse my lack of knowledge here, but how do I specify this? I am unable to find the relevant documentation or an example.

I tried to simply use a `resources` folder in the template but this just copied the entire resources folder (including the hidden files though). Obviously I need the files in the base directory of the project.

**Reply by mcanouil** [reply]
![image](https://github.com/quarto-dev/quarto-cli/assets/8896044/80f803a1-b88a-4894-a1e9-fffc0031ba49)
See <https://quarto.org/docs/extensions/formats.html#format-resources>.

For `resources`, it's a website/book project level option (a bit tricker to find): https://quarto.org/docs/websites/website-tools.html#site-resources

**Reply by mcanouil** [reply]
By the way, why make the `.devcontainer` inside the extension? use `devcontainer` and set some workflow in your extension to use that in the root of your project as `.devcontainer`.

VSCode won't look for `_extensions/<account>/<repo>/.devcontainer` so I don't see the rational for this one at least.

**Reply by skwde** [reply]
Ok, I am lost now.

Here is what I actually would like to achieve:

```sh
quarto use template my-template
```

This should set up a project folder with following content:

```out
├── .devcontainer
│   └── devcontainer.json
├── .gitignore
├── .pre-commit-config.yaml
├── Dockerfile
└── template.qmd
```

Consequently `my-template` should contain all this files, but I am unsure about where to put them.
The created project folder is tracked via git and available on a private gitlab instance.

Currently I always have to copy the .devcontainer and other hidden files by hand.

Based on your last comment I have the impression that using `_extensions` is not the way to achieve that.

How do I create a new project with hidden files?

@mcanouil, I edited the post to better reflect what I actually try to achieve and make it clearer why I would like to do that. Besides that I am still confused about the difference of extension / template.

**Reply by mcanouil** [reply]
The following in an extension will copy the defined format-resources into the root of your Quarto project (https://quarto.org/docs/extensions/formats.html#format-resources):

```yaml
contributes:
  format:
    html:
      format-resources:
        - .devcontainer
```

Note that `.devcontainer` has to be inside the extension:

```md
_extensions
└── NAME
    ├── .devcontainer
    │   └── devcontainer.json 
    └── _extension.yml
```

Edit: It actually does not work exactly as expected, as `format-resources` will only copy the resources prior to rendering.

**Reply by skwde** [reply]
Ok, thanks for your reply.

So just to confirm, it is currently not possible to create a project directory from a skeleton. 
I simply want to have a custom project files (project not in the quarto sense, but in the git repository sense).

The approach you mention seems like a workaround to what I try to achieve. (I don't want to change rendering / preview behavior or ship my own css style).
From what I understand the use of *template*s is not meant for that.
I guess I wrongly understood quarto projects / templates in the first place.

I however thing that something like this is quite useful to more than me, should this move to a feature request?

**Reply by mcanouil** [reply]
> project not in the quarto sense, but in the git repository sense

Then use Git/GitHub (or GitLab) and create a repository template.

**Reply by skwde** [reply]
Yes that's what I try to do, `quarto use template` allows for git repositories

When using `git clone` for my template, another another manual interaction is required, namely removing `.git`.
That is why I try to use quarto templates in the first place.

Are quarto templates only meant to be used with and `_extension` folder?
I don't quite get why there is the restriction of not copying any hidden files, at least it should be optin, e.g. with an option.
Point 1 and 2 at https://quarto.org/docs/extensions/starter-templates.html even mention it, that it is meant for initial content.

**Reply by mcanouil** [reply]
> Yes that's what I try to do, quarto use template allows for git repositories

A "template" is a format extension which contains a `template.qmd`.

There are no restrictions. It's a current limitation as there is nothing to tell Quarto what files/directories should be kept or not from the `_extensions` directory in the Git repository.

Extensions were designed to allow users to use whatever custom formats they want because custom formats require repeated and tedious tasks without.


FYI:
- https://github.com/quarto-dev/quarto-cli/issues/9515

> with an option.
Point 1 and 2 at https://quarto.org/docs/extensions/starter-templates.html even mention it

You have to read the links to understand what "initial content" means.

**Reply by skwde** [reply]
Thanks again for your response and explanation.

The issue is indeed helpful.

Now patiently waiting for it to be solved ;)

**Comment by NeuroShepherd** [comment]
I was not able to accomplish this with any combination of including files via `resources`, `format-resources,` or specifying exact or glob paths to dot files in my `.quartoignore` like the original poster.

However, I found a (hacky) solution to this where I first negated the exclusion of all dot files/folders, and then added back the paths I wanted to ignore. E.g. a testing project of mine is structured like this:

```
tree -La 1
.
├── _extensions
├── _quarto.yml
├── _site
├── .DS_Store
├── .git
├── .github
├── .gitignore
├── .posit
├── .quarto
├── .quartoignore
├── apage.qmd
├── index.qmd
├── README.md
└── template.qmd
```

and my `.quartoignore` file contains this:

```
!.*
.git/
.quarto/
.posit/
.quartoignore
```

which copies over my`.github` and `.gitignore` while ignoring the other dot files.

I have a suspicion there's an issue (or at least a misunderstanding from the user-side) with the [`template.ts`](https://github.com/quarto-dev/quarto-cli/blob/main/src/extension/template.ts) implementation. It seems this code:

https://github.com/quarto-dev/quarto-cli/blob/effe14586db57a9e09271ab9ac4ed35fcbd300dc/src/extension/template.ts#L37-L41

filters on exact string matches, which means a broad exclusion pattern like ".*" from kBuiltInExcludes will *not* be removed by a more specific include pattern like ".github/**" because they are not identical strings, leading to the broader exclusion taking precedence.

**Comment by mcanouil** [comment]
Using Quarto Wizard for Positron/VSCode, you can choose whatever is available inside the Quarto extension directory/repository, including hidden directories/files.

See https://m.canouil.dev/quarto-wizard/getting-started/using-templates.html#installing-a-template

**Reply by NeuroShepherd** [reply]
Neat, thanks for the additional work-around!
